compiler/rustc_errors/src/emitter.rs

```diff
@@ -2302,6 +2367,201 @@ impl<'a> Drop for WritableDst<'a> {
     }
 }
 
+/// A type that formats everything it receives to HTML. It implements both
+/// Write (so that we can write text in it) and WriteColor (so that we can
+/// set which color to use).
+pub struct HtmlFormatter {
+    inner: Box<dyn Write>,
+}
+
+impl HtmlFormatter {
+    fn new(inner: Box<dyn Write>) -> HtmlFormatter {
+        HtmlFormatter { inner }
+    }
+
+    fn begin(&mut self) -> io::Result<()> {
+        write!(self.inner, r#"<pre style="margin:0px">"#)
+    }
+
+    fn end(&mut self) -> io::Result<()> {
+        write!(self.inner, r#"</pre>"#)
+    }
+
+    // Creates the string of a `style` attribute in an html element
+    fn mk_style_string(spec: &ColorSpec) -> String {
+        let mut buffer = String::new();
+
+        let colors = [("color", spec.fg()), ("background-color", spec.bg())];
+        let colors = colors.iter().flat_map(|(c_name, c_value)| {
+            c_value.map(|c_value| (c_name, term_color_to_html_color(*c_value)))
+        });
+
+        for (key, value) in colors {
+            buffer.push_str(format!("{}: {};", key, value).as_str());
+        }
+
+        let font_modifiers = [
+            ("font-weight:bold", spec.bold()),
+            ("font-style:italic", spec.italic()),
+            ("text-decoration:underline", spec.underline()),
+        ];
+        let font_modifiers = font_modifiers
+            .iter()
+            .flat_map(|(property, applicable)| if *applicable { Some(property) } else { None })
+            .copied();
+
+        for font_modifier in font_modifiers {
+            buffer.push_str(format!("{};", font_modifier).as_str());
+        }
+
+        // The trailing semicolon must removed as defined in the style
+        // attribute grammar:
+        // https://drafts.csswg.org/css-style-attr/#syntax
+        let _ = buffer.pop();
+
+        buffer
+    }
+}
+
+impl Write for HtmlFormatter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        for (idx, segment) in EscapedHtmlIter::new(buf) {
+            match segment {
+                EscapedHtmlSegment::Modified(buf) => match self.inner.write_all(buf) {
+                    Ok(()) => {}
+
+                    Err(e) if e.kind() == ErrorKind::WriteZero => return Ok(idx),
+                    Err(e) => return Err(e),
```

Comment:
In practice I don't know if it actually matters (I am assuming that you've run tests and observed it work; I haven't tried to work on the compiler before and so haven't actually downloaded or run this myself), but I *think* this clause loses the current `.next()` segment if it is not able to write into `self.inner`, so a retry would skip it. If it works as-is, then I wouldn't worry about it.

Reply:
Actually it is not a problem if we don't write the current `.next()`, as long as we exit early and return the correct amount of bytes we actually wrote.

But the good news is that you're highlighting the fact that we should not return an error if we wrote something to the sink. I do need to fix that.